services/dynamic-scheduler/tests/unit/service_tracker/test__api.py

```diff
@@ -0,0 +1,321 @@
+            SchedulerServiceState.STARTING,
+        ),
+        (
+            UserRequestedState.RUNNING,
+            status_generator(ServiceState.STARTING),
+            SchedulerServiceState.STARTING,
+        ),
+        (
+            UserRequestedState.RUNNING,
+            status_generator(ServiceState.RUNNING),
+            SchedulerServiceState.RUNNING,
+        ),
+        (
+            UserRequestedState.RUNNING,
+            status_generator(ServiceState.COMPLETE),
+            SchedulerServiceState.UNEXPECTED_OUTCOME,
+        ),
+        (
+            UserRequestedState.RUNNING,
+            status_generator(ServiceState.FAILED),
+            SchedulerServiceState.UNEXPECTED_OUTCOME,
+        ),
+        (
+            UserRequestedState.RUNNING,
+            status_generator(ServiceState.STOPPING),
+            SchedulerServiceState.UNEXPECTED_OUTCOME,
+        ),
+        (
+            UserRequestedState.RUNNING,
+            _get_node_get_idle(),
+            SchedulerServiceState.IDLE,
+        ),
+        # UserRequestedState.STOPPED
+        (
+            UserRequestedState.STOPPED,
+            status_generator(ServiceState.PENDING),
+            SchedulerServiceState.UNEXPECTED_OUTCOME,
+        ),
+        (
+            UserRequestedState.STOPPED,
+            status_generator(ServiceState.PULLING),
+            SchedulerServiceState.UNEXPECTED_OUTCOME,
+        ),
+        (
+            UserRequestedState.STOPPED,
+            status_generator(ServiceState.STARTING),
+            SchedulerServiceState.UNEXPECTED_OUTCOME,
+        ),
+        (
+            UserRequestedState.STOPPED,
+            status_generator(ServiceState.RUNNING),
+            SchedulerServiceState.STOPPING,
+        ),
+        (
+            UserRequestedState.STOPPED,
+            status_generator(ServiceState.COMPLETE),
+            SchedulerServiceState.STOPPING,
+        ),
+        (
+            UserRequestedState.STOPPED,
+            status_generator(ServiceState.FAILED),
+            SchedulerServiceState.UNEXPECTED_OUTCOME,
+        ),
+        (
+            UserRequestedState.STOPPED,
+            status_generator(ServiceState.STOPPING),
+            SchedulerServiceState.STOPPING,
+        ),
+        (
+            UserRequestedState.STOPPED,
+            _get_node_get_idle(),
+            SchedulerServiceState.IDLE,
+        ),
+    ]
+    for status_generator in (
+        _get_node_get_from,
+        _get_dynamic_service_get_from,
+    )
+]
+_FLAT_EXPECTED_TEST_CASES = __get_flat_list(_EXPECTED_TEST_CASES)
+# ensure enum changes do not break above rules
+_IDLE_ITEM_COUNT: Final[int] = 1
+_NODE_STATUS_FORMATS_COUNT: Final[int] = 2
+assert (
+    len(_FLAT_EXPECTED_TEST_CASES)
+    == (len(ServiceState) + _IDLE_ITEM_COUNT)
+    * len(UserRequestedState)
+    * _NODE_STATUS_FORMATS_COUNT
+)
+
+
+@pytest.mark.parametrize("requested_state, status, expected", _FLAT_EXPECTED_TEST_CASES)
```

Comment:
maybe create a dataclass for this parameter. it would be easier to read

Reply:
I guess you refer to group all three parameters. I can do that. Let me know if the name is strange.